be2-scala/src/main/scala/ch/epfl/pop/storage/DbActorNew.scala

```diff
@@ -34,24 +34,88 @@ case class DbActorNew(
     }
     createChannel(channel, _object)
 
-    val channelData: ChannelData = readChannelData(channel)
-    storage.write(
-      (channel.toString, channelData.addMessage(message.message_id).toJsonString),
-      (s"$channel:${message.message_id}", message.toJsonString)
-    )
+    this.synchronized{
+      val channelData: ChannelData = readChannelData(channel)
+      storage.write(
+        (channel.toString, channelData.addMessage(message.message_id).toJsonString),
+        (s"$channel${Channel.DATA_SEPARATOR}${message.message_id}", message.toJsonString)
+      )
+    }
   }
 
   @throws [DbActorNAckException]
-  private def read(channel: Channel, messageId: Hash): Option[Message] = ???
+  private def read(channel: Channel, messageId: Hash): Option[Message] = {
+    Try(storage.read(s"$channel${Channel.DATA_SEPARATOR}$messageId")) match {
+      case Success(Some(json)) => Some(Message.buildFromJson(json))
+      case Success(None) => None
+      case Failure(ex) => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, ex.getMessage)
+    }
+  }
 
   @throws [DbActorNAckException]
-  private def readChannelData(channel: Channel): ChannelData = ???
+  private def readChannelData(channel: Channel): ChannelData = {
+    Try(storage.read(channel.toString)) match {
+      case Success(Some(json)) => ChannelData.buildFromJson(json)
+      case Success(None) => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, s"ChannelData for channel $channel not in the database")
+      case Failure(ex) => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, ex.getMessage)
+    }
+  }
 
   @throws [DbActorNAckException]
-  private def readLaoData(channel: Channel): LaoData = ???
+  private def readLaoData(channel: Channel): LaoData = {
+    Try(storage.read(generateLaoDataKey(channel))) match {
+      case Success(Some(json)) => LaoData.buildFromJson(json)
+      case Success(None) => LaoData.emptyLaoData
+      case Failure(ex) => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, ex.getMessage)
+    }
+  }
 
   @throws [DbActorNAckException]
-  private def catchupChannel(channel: Channel): List[Message] = ???
+  private def writeLaoData(channel: Channel, message: Message): Unit = {
+    this.synchronized{
+      val laoData: LaoData = Try(readLaoData(channel)) match {
+        case Success(data) => data
+        case Failure(_) => LaoData.emptyLaoData
+      }
+      val laoDataKey: String = generateLaoDataKey(channel)
+
+      Try(storage.write(
+        (laoDataKey, laoData.updateWith(message).toJsonString)
+      )).recover(
+        ex => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, ex.getMessage)
+      )
```

Comment:
No need to recover since you are throwing the same thing :)

```suggestion
      storage.write(laoDataKey -> laoData.updateWith(message).toJsonString)
```

Reply:
Ohhhh I now understand why you had it like that... It's because I did the same mistake ^^' Since the DiskStorage only fails with dbactornackexceptions, we don't need these recover anymore :D